### ISSUE TITLE
Using a list of extended trait names in @on_trait_change gives obscure TraitError

### ISSUE THREAD

**Issue by mdickinson** [enthought/traits#403]
When I run the code shown below, I get the following traceback:
```python
Traceback (most recent call last):
  File "traits_ex.py", line 21, in <module>
    model = B()
  File "/Users/mdickinson/.edm/envs/traits-futures-py36-pyqt/lib/python3.6/site-packages/traits/has_traits.py", line 3329, in _init_trait_listeners
    getattr( self, '_init_trait_%s_listener' % data[0] )( name, *data )
  File "/Users/mdickinson/.edm/envs/traits-futures-py36-pyqt/lib/python3.6/site-packages/traits/has_traits.py", line 3339, in _init_trait_method_listener
    dispatch=config['dispatch'] )
  File "/Users/mdickinson/.edm/envs/traits-futures-py36-pyqt/lib/python3.6/site-packages/traits/has_traits.py", line 2575, in on_trait_change
    priority, target )
  File "/Users/mdickinson/.edm/envs/traits-futures-py36-pyqt/lib/python3.6/site-packages/traits/has_traits.py", line 2612, in on_trait_change
    deferred        = deferred )
  File "/Users/mdickinson/.edm/envs/traits-futures-py36-pyqt/lib/python3.6/site-packages/traits/has_traits.py", line 1543, in trait_set
    setattr( self, name, value )
  File "/Users/mdickinson/.edm/envs/traits-futures-py36-pyqt/lib/python3.6/site-packages/traits/trait_handlers.py", line 172, in error
    value )
traits.trait_errors.TraitError: The 'deferred' trait of a ListenerItem instance must be a boolean, but a value of None <class 'NoneType'> was specified.
```

This may just be a documentation bug, as it's not clear from the documentation whether this should be permitted.

Versions: Python 3.6.0; Traits 4.6.0; macOS 10.13.5.

Example code:
```python
from traits.api import HasStrictTraits, Instance, Int, Unicode, on_trait_change


class A(HasStrictTraits):
    foo = Unicode()

    bar = Int()


class B(HasStrictTraits):
    a_instance = Instance(A)

    something_else = Int()

    @on_trait_change(['something_else', 'a_instance:foo', 'a_instance:bar'])
    def react_to_change(self):
        print("Something changed")


if __name__ == '__main__':
    model = B()
```

**Comment by corranwebster** [comment]
I don't think that this is supposed to work, and I have seen it as a gotcha before, so it would be good to fix to either raise or work as expected.

**Comment by mdickinson** [comment]
So the docstring for `HasTraits.on_trait_change` says:
```
        The *name* parameter is a single *xname* or a list of *xname* names,
        where an *xname* is an extended name of the form::
```
... which suggests that this should work.

**Comment by mdickinson** [comment]
The error seems to be the result of a simple mismatch between the `on_trait_change` [signature](https://github.com/enthought/traits/blob/a99b3f64d50c5f7f28ffc01bf69419b061f9e976/traits/has_traits.py#L2341-L2343) and an `on_trait_change` [call](https://github.com/enthought/traits/blob/a99b3f64d50c5f7f28ffc01bf69419b061f9e976/traits/has_traits.py#L2565-L2566): the `target` actual value is bound to the `deferred` formal parameter name.

**Comment by mdickinson** [comment]
Fixed in #404.
